File: java/PlanesAndroid/app/src/main/java/com/planes/android/PlanesVerticalLayout.java
package com.planes.android;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.HashMap;


//vertical layout works with 2 GameBoard objects and 2 GameControl objects
//the game control object should change depending on the game stage
//the size of the game controls depends on the size of the visible game board
//dimension of the screen and toolbars should be saved inside the layout
public class PlanesVerticalLayout extends ViewGroup {

    class PlanesVerticalLayoutParams extends LayoutParams {

        PlanesVerticalLayoutParams(Context context, AttributeSet attrs) {
            super(context, attrs);
            TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.PlanesVerticalLayout);
            m_Row = a.getInt(R.styleable.PlanesVerticalLayout_gc_row, 0);
            m_Col = a.getInt(R.styleable.PlanesVerticalLayout_gc_col, 0);
            m_GameStage = a.getInt(R.styleable.PlanesVerticalLayout_gc_game_stage, 0);
            a.recycle();
        }

        PlanesVerticalLayoutParams(int width, int height) {
            super(width, height);
        }

        PlanesVerticalLayoutParams(ViewGroup.LayoutParams source) {
            super(source);
        }

        private int m_Row = 0;
        private int m_Col = 0;
        private int m_RowSpan = 0;
        private int m_ColSpan = 0;
        private int m_GameStage = -1;
        private boolean m_Player = false;
    }

    public PlanesVerticalLayout(Context context) {
        super(context);
        m_Context = context;
    }

    public PlanesVerticalLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        m_Context = context;
    }

    public PlanesVerticalLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        m_Context = context;
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        final int count = getChildCount();

        m_GameBoards = new ArrayList<GameBoard>();
        m_GameControls = new HashMap<Integer, ArrayList<View>>();
        m_GameControlsMaxRow = new HashMap<Integer, Integer>();
        m_GameControlsMaxCol = new HashMap<Integer, Integer>();

        //TODO: to move in a init function
        //TODO: to read only the children corresponding to the current stage
        for (int i = 0; i < count; i++) {
            View child = getChildAt(i);

            if (child instanceof GameBoard) {
                m_GameBoards.add((GameBoard)child);
            } else {
                PlanesVerticalLayoutParams lp = (PlanesVerticalLayoutParams) child.getLayoutParams();

                if (!(lp.m_Row == 0 || lp.m_Col == 0 || lp.m_GameStage == -1)) {
                    if (!m_GameControls.containsKey(lp.m_GameStage)) {
                        m_GameControls.put(lp.m_GameStage, new ArrayList<View>());
                        m_GameControlsMaxRow.put(lp.m_GameStage, 0);
                        m_GameControlsMaxCol.put(lp.m_GameStage, 0);
                    }
                    ArrayList<View> viewsForGameStage = m_GameControls.get(lp.m_GameStage);
                    if (lp.m_GameStage != m_GameStage.getValue()) {
                        child.setVisibility(GONE);
                    } else {
                        child.setVisibility(VISIBLE);
                    }
                    viewsForGameStage.add(child);
                    int maxRow = m_GameControlsMaxRow.get(lp.m_GameStage);
                    if (lp.m_Row > maxRow)
                        m_GameControlsMaxRow.put(lp.m_GameStage, lp.m_Row);
                    int maxCol = m_GameControlsMaxCol.get(lp.m_GameStage);
                    if (lp.m_Col > maxCol)
                        m_GameControlsMaxCol.put(lp.m_GameStage, lp.m_Col);
                }
            }
        }

        if (m_GameBoards.size() == 0 || m_GameBoards.size() > 2) {
            return;
        }

        if (m_GameBoards.size() == 2) {
            m_Tablet = true;
        }

        if (right - left < bottom - top) { //vertical layout
            m_Vertical = true;
        }

        int layoutWidth = right -left;
        int layoutHeight = bottom - top;

        if (m_GameStage == GameStages.BoardEditing) {
            if (m_Tablet) {
                if (m_Vertical) {
                    setPlayerBoardPosition(0, 0, layoutWidth,layoutHeight / 2, true);
                    setGameControlsPositions(0, layoutHeight / 2, layoutWidth, layoutHeight);
                } else {
                    setPlayerBoardPosition(0, 0,layoutWidth / 2,  layoutHeight, true);
                    setGameControlsPositions(layoutWidth / 2, 0, layoutWidth, layoutHeight);
                }
            } else {
                if (m_Vertical) {
                    setFirstBoardPosition(0, 0, layoutWidth, layoutWidth);
                    setGameControlsPositions(0, layoutWidth, layoutWidth, layoutHeight);
                } else {
                    setFirstBoardPosition(0, 0, layoutHeight, layoutHeight);
                    setGameControlsPositions(layoutHeight, 0, layoutWidth, layoutHeight);
                }
            }
        }

        if (m_GameStage == GameStages.Game) {
            if (m_Tablet) {
                if (m_Vertical) {
                    setPlayerBoardPosition(0, 0, layoutWidth,layoutHeight / 2, false);
                    setComputerBoardPosition(0, layoutHeight / 2, layoutWidth, layoutHeight, false);
                } else {
                    setPlayerBoardPosition(0, 0,layoutWidth / 2,  layoutHeight, false);
                    setComputerBoardPosition(layoutWidth / 2, 0, layoutWidth, layoutHeight, false);
                }
            } else {
                if (m_Vertical) {
                    setFirstBoardPosition(0, 0, layoutWidth, layoutWidth);
                    setGameControlsPositions(0, layoutWidth, layoutWidth, layoutHeight);
                } else {
                    setFirstBoardPosition(0, 0, layoutHeight, layoutHeight);
                    setGameControlsPositions(layoutHeight, 0, layoutWidth, layoutHeight);
                }
            }
        }

        if (m_GameStage == GameStages.GameNotStarted) {
            if (m_Tablet) {
                if (m_Vertical) {
                    if (m_ShowPlayerBoard) {
                        setPlayerBoardPosition(0, 0, layoutWidth,layoutHeight / 2, true);
                        setGameControlsPositions(0, layoutHeight / 2, layoutWidth, layoutHeight);
                    } else {
                        setGameControlsPositions(0, 0, layoutWidth,layoutHeight / 2);
                        setComputerBoardPosition(0, layoutHeight / 2, layoutWidth, layoutHeight, true);
                    }
                } else {
                    if (m_ShowPlayerBoard) {
                        setGameControlsPositions(0, 0,layoutWidth / 2,  layoutHeight);
                        setPlayerBoardPosition(layoutWidth / 2, 0, layoutWidth, layoutHeight, true);
                    } else {
                        setComputerBoardPosition(0, 0,layoutWidth / 2,  layoutHeight, true);
                        setGameControlsPositions(layoutWidth / 2, 0, layoutWidth, layoutHeight);
                    }
                }
            } else {
                if (m_Vertical) {
                    setFirstBoardPosition(0, 0, layoutWidth, layoutWidth);
                    setGameControlsPositions(0, layoutWidth, layoutWidth, layoutHeight);
                } else {
                    setFirstBoardPosition(0, 0, layoutHeight, layoutHeight);
                    setGameControlsPositions(layoutHeight, 0, layoutWidth, layoutHeight);
                }
            }
        }
    }

    @Override
    public LayoutParams generateLayoutParams(AttributeSet attrs) {
        return new PlanesVerticalLayoutParams(getContext(), attrs);
    }
    @Override
    protected LayoutParams generateDefaultLayoutParams() {
        return new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
    }
    @Override
    protected ViewGroup.LayoutParams generateLayoutParams(ViewGroup.LayoutParams p) {
        return new LayoutParams(p);
    }
    @Override
    protected boolean checkLayoutParams(ViewGroup.LayoutParams p) {
        return p instanceof LayoutParams;
    }


    private void setPlayerBoardPosition(int left, int top, int right, int bottom, boolean hideOthers) {
        for (GameBoard board : m_GameBoards) {
            if (board.isPlayer()) {
                board.layout(left, top, right, bottom);
                board.setVisibility(VISIBLE);
            } else {
                if (hideOthers)
                    board.setVisibility(GONE);
                else
                    board.setVisibility(VISIBLE);
            }
        }
    }

    private void setComputerBoardPosition(int left, int top, int right, int bottom, boolean hideOthers) {
        for (GameBoard board : m_GameBoards) {
            if (!board.isPlayer()) {
                board.layout(left, top, right, bottom);
                board.setVisibility(VISIBLE);
            } else {
                if (hideOthers)
                    board.setVisibility(GONE);
                else
                    board.setVisibility(VISIBLE);
            }
        }
    }

    private void setFirstBoardPosition(int left, int top, int right, int bottom) {
        m_GameBoards.get(0).layout(left, top,  right, bottom);
    }

    private void setGameControlsPositions(int left, int top, int right, int bottom) {
        int maxRow = m_GameControlsMaxRow.get(m_GameStage.getValue());
        int maxCol = m_GameControlsMaxCol.get(m_GameStage.getValue());

        int stepX = (right - left) / (maxCol + 2);
        int stepY = (bottom - top) / (maxRow + 2);

        for (View view : m_GameControls.get(m_GameStage.getValue())) {
            PlanesVerticalLayoutParams lp = (PlanesVerticalLayoutParams) view.getLayoutParams();

            //TODO: create own buttons and textviews
            /*int measWidth = view.getMeasuredWidth();
            int measHeight = view.getMeasuredHeight();

            int actualWidth = Math.max(stepX, measWidth);
            int actualHeight = Math.max(stepY, measHeight);

            int viewCenterX = left + lp.m_Col * stepX + stepX / 2;
            int viewCenterY = top + lp.m_Row * stepY + stepY / 2;

            view.layout(viewCenterX - actualWidth / 2, viewCenterY - actualHeight / 2, viewCenterX + actualWidth / 2, viewCenterY + actualWidth / 2);*/

            view.layout(left + lp.m_Col * stepX, top + lp.m_Row * stepY, left + (lp.m_Col + 1) * stepX, top + (lp.m_Row + 1) * stepY);
        }
    }

    //TODO: is the use of invalidate appropriate?
    public void setNewRoundStage() {
        m_GameStage = GameStages.GameNotStarted;
        invalidate();
    }

    public void setGameStage() {
        m_GameStage = GameStages.Game;
        invalidate();
        requestLayout();
    }

    public void setBoardEditingStage() {
        m_GameStage = GameStages.BoardEditing;
        invalidate();
        requestLayout();
    }

    public void setPlayerBoard() {
        m_ShowPlayerBoard = true;
        invalidate();
        requestLayout();
    }

    public void setComputerBoard() {
        m_ShowPlayerBoard = false;
        invalidate();
        requestLayout();
    }


    private ArrayList<GameBoard> m_GameBoards;
    private HashMap<Integer, ArrayList<View>> m_GameControls;
    private HashMap<Integer, Integer> m_GameControlsMaxRow;
    private HashMap<Integer, Integer> m_GameControlsMaxCol;

    private Context m_Context;
    private GameStages m_GameStage = GameStages.BoardEditing;
    private boolean m_CorrectChildren;

    private boolean m_ShowPlayerBoard = false;  //in start new game stage which board to show
    private boolean m_Tablet = false;
    private boolean m_Vertical = false;
}

File: kotlin/PlanesAndroid/app/src/main/java/com/planes_multiplayer/android/PlanesVerticalLayout.kt
package com.planes_multiplayer.android

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import java.util.*

//vertical layout works with 2 GameBoard objects and 2 GameControl objects
//the game control object should change depending on the game stage
//the size of the game controls depends on the size of the visible game board
//dimension of the screen and toolbars should be saved inside the layout
class PlanesVerticalLayout : ViewGroup {
    internal inner class PlanesVerticalLayoutParams : LayoutParams {
        constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
            val a = context.obtainStyledAttributes(attrs, R.styleable.PlanesVerticalLayout)
            m_Row = a.getInt(R.styleable.PlanesVerticalLayout_gc_row, 0)
            m_Col = a.getInt(R.styleable.PlanesVerticalLayout_gc_col, 0)
            m_RowSpan = a.getInt(R.styleable.PlanesVerticalLayout_gc_rowspan, 0)
            m_ColSpan = a.getInt(R.styleable.PlanesVerticalLayout_gc_colspan, 0)
            m_GameStage = a.getInt(R.styleable.PlanesVerticalLayout_gc_game_stage, 0)
            m_Text = a.getString(R.styleable.PlanesVerticalLayout_gc_text).toString()
            m_Text1 = a.getString(R.styleable.PlanesVerticalLayout_gc_text1).toString()
            m_Text2 = a.getString(R.styleable.PlanesVerticalLayout_gc_text2).toString()
            m_BackgroundColor = a.getColor(R.styleable.PlanesVerticalLayout_gc_background_color, resources.getColor(R.color.grey))
            a.recycle()
        }

        //TODO: for these 2 constructors there is no initialization of member variables !!!
        constructor(width: Int, height: Int) : super(width, height) {}
        constructor(source: LayoutParams?) : super(source) {}

        fun getText(): String {
            return m_Text
        }

        fun getText1(): String {
            return m_Text1
        }

        fun getText2(): String {
            return m_Text2
        }

        fun getColor(): Int {
            return m_BackgroundColor
        }

        //TODO: to add text formatting options
        //word wrap, no word wrap
        var m_Row = 0
        var m_Col = 0
        var m_RowSpan = 0
        var m_ColSpan = 0
        var m_GameStage = -1
        private val m_Player = false
        lateinit var m_Text: String
        lateinit var m_Text1: String
        lateinit var m_Text2: String
        var m_BackgroundColor = 0
    }

    constructor(context: Context) : super(context) {
        m_Context = context
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        m_Context = context
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        m_Context = context
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        val count = childCount
        m_GameBoards = ArrayList()
        m_GameControls = HashMap()
        m_GameControlsMaxRow = HashMap()
        m_GameControlsMaxCol = HashMap()

        //TODO: to move in a init function
        //TODO: to read only the children corresponding to the current stage
        for (i in 0 until count) {
            val child = getChildAt(i)
            if (child is GameBoard) {
                m_GameBoards.add(child)
            } else {
                val lp = child.layoutParams as PlanesVerticalLayoutParams
                if (!(lp.m_Row == 0 || lp.m_Col == 0 || lp.m_GameStage == -1)) {
                    if (!m_GameControls.containsKey(lp.m_GameStage)) {
                        m_GameControls[lp.m_GameStage] = ArrayList()
                        m_GameControlsMaxRow[lp.m_GameStage] = 0
                        m_GameControlsMaxCol[lp.m_GameStage] = 0
                    }
                    val viewsForGameStage = m_GameControls[lp.m_GameStage]!!
                    if (lp.m_GameStage != m_GameStage.value) {
                        child.visibility = GONE
                    } else {
                        child.visibility = VISIBLE
                    }
                    viewsForGameStage.add(child)
                    val maxRow = m_GameControlsMaxRow[lp.m_GameStage]!!
                    val rowspan = if (lp.m_RowSpan != 0) lp.m_RowSpan - 1 else 0
                    val colspan = if (lp.m_ColSpan != 0) lp.m_ColSpan - 1 else 0
                    if (lp.m_Row + rowspan > maxRow) m_GameControlsMaxRow[lp.m_GameStage] = lp.m_Row + rowspan
                    val maxCol = m_GameControlsMaxCol[lp.m_GameStage]!!
                    if (lp.m_Col + colspan > maxCol) m_GameControlsMaxCol[lp.m_GameStage] = lp.m_Col + colspan
                }
            }
        }
        if (m_GameBoards.size == 0 || m_GameBoards.size > 2) {
            return
        }
        if (m_GameBoards.size == 2) {
            m_Tablet = true
        }
        if (right - left < bottom - top) { //vertical layout
            m_Vertical = true
        }
        val layoutWidth = right - left
        val layoutHeight = bottom - top
        if (m_GameStage === GameStages.BoardEditing) {
            if (m_Tablet) {
                if (m_Vertical) {
                    setPlayerBoardPosition(0, 0, layoutWidth, layoutHeight / 2, true)
                    setGameControlsPositions(0, layoutHeight / 2, layoutWidth, layoutHeight)
                } else {
                    setPlayerBoardPosition(0, 0, layoutWidth / 2, layoutHeight, true)
                    setGameControlsPositions(layoutWidth / 2, 0, layoutWidth, layoutHeight)
                }
            } else {
                if (m_Vertical) {
                    setFirstBoardPosition(0, 0, layoutWidth, layoutWidth)
                    setGameControlsPositions(0, layoutWidth, layoutWidth, layoutHeight)
                } else {
                    setFirstBoardPosition(0, 0, layoutHeight, layoutHeight)
                    setGameControlsPositions(layoutHeight, 0, layoutWidth, layoutHeight)
                }
            }
        }
        val boardSpacing = 20
        if (m_GameStage === GameStages.Game) {
            if (m_Tablet) {
                if (m_Vertical) {
                    setPlayerBoardPosition(0, 0, layoutWidth, layoutHeight / 2 - boardSpacing, false)
                    setComputerBoardPosition(0, layoutHeight / 2 + boardSpacing, layoutWidth, layoutHeight, false)
                } else {
                    setPlayerBoardPosition(0, 0, layoutWidth / 2 - boardSpacing, layoutHeight, false)
                    setComputerBoardPosition(layoutWidth / 2 + boardSpacing, 0, layoutWidth, layoutHeight, false)
                }
            } else {
                if (m_Vertical) {
                    setFirstBoardPosition(0, 0, layoutWidth, layoutWidth)
                    setGameControlsPositions(0, layoutWidth, layoutWidth, layoutHeight)
                } else {
                    setFirstBoardPosition(0, 0, layoutHeight, layoutHeight)
                    setGameControlsPositions(layoutHeight, 0, layoutWidth, layoutHeight)
                }
            }
        }
        if (m_GameStage === GameStages.GameNotStarted) {
            if (m_Tablet) {
                if (m_Vertical) {
                    if (!m_ShowPlayerBoard) {
                        setGameControlsPositions(0, 0, layoutWidth, layoutHeight / 2)
                        setComputerBoardPosition(0, layoutHeight / 2, layoutWidth, layoutHeight, true)
                    } else {
                        setPlayerBoardPosition(0, 0, layoutWidth, layoutHeight / 2, true)
                        setGameControlsPositions(0, layoutHeight / 2, layoutWidth, layoutHeight)
                    }
                } else {
                    if (m_ShowPlayerBoard) {
                        setPlayerBoardPosition(0, 0, layoutWidth / 2, layoutHeight, true)
                        setGameControlsPositions(layoutWidth / 2, 0, layoutWidth, layoutHeight)
                    } else {
                        setGameControlsPositions(0, 0, layoutWidth / 2, layoutHeight)
                        setComputerBoardPosition(layoutWidth / 2, 0, layoutWidth, layoutHeight, true)
                    }
                }
            } else {
                if (m_Vertical) {
                    setFirstBoardPosition(0, 0, layoutWidth, layoutWidth)
                    setGameControlsPositions(0, layoutWidth, layoutWidth, layoutHeight)
                } else {
                    setFirstBoardPosition(0, 0, layoutHeight, layoutHeight)
                    setGameControlsPositions(layoutHeight, 0, layoutWidth, layoutHeight)
                }
            }
        }
    }

    override fun generateLayoutParams(attrs: AttributeSet): LayoutParams {
        return PlanesVerticalLayoutParams(context, attrs)
    }

    override fun generateDefaultLayoutParams(): LayoutParams {
        return LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    }

    override fun generateLayoutParams(p: LayoutParams): LayoutParams {
        return LayoutParams(p)
    }

    override fun checkLayoutParams(p: LayoutParams): Boolean {
        return p is LayoutParams
    }

    private fun setPlayerBoardPosition(left: Int, top: Int, right: Int, bottom: Int, hideOthers: Boolean) {
        for (board in m_GameBoards) {
            if (board.isPlayer()) {
                board.layout(left, top, right, bottom)
                board.visibility = VISIBLE
            } else {
                if (hideOthers) board.visibility = GONE else board.visibility = VISIBLE
            }
        }
    }

    private fun setComputerBoardPosition(left: Int, top: Int, right: Int, bottom: Int, hideOthers: Boolean) {
        for (board in m_GameBoards) {
            if (!board.isPlayer()) {
                board.layout(left, top, right, bottom)
                board.visibility = VISIBLE
            } else {
                if (hideOthers) board.visibility = GONE else board.visibility = VISIBLE
            }
        }
    }

    private fun setFirstBoardPosition(left: Int, top: Int, right: Int, bottom: Int) {
        m_GameBoards[0]!!.layout(left, top, right, bottom)
    }

    private fun setGameControlsPositions(left: Int, top: Int, right: Int, bottom: Int) {
        val maxRow = m_GameControlsMaxRow[m_GameStage.value]!!
        val maxCol = m_GameControlsMaxCol[m_GameStage.value]!!
        val stepX = (right - left) / (maxCol + 2)
        val stepY = (bottom - top) / (maxRow + 2)
        var currentOptimalTextSize = 100

        //compute text size
        for (view in m_GameControls[m_GameStage.value]!!) {
            if (view !is ViewWithText) continue
            val lp = view.layoutParams as PlanesVerticalLayoutParams
            val rowspan = if (lp.m_RowSpan != 0) lp.m_RowSpan else 1
            val colspan = if (lp.m_ColSpan != 0) lp.m_ColSpan else 1
            val heightMeasureSpec = MeasureSpec.makeMeasureSpec(rowspan * stepY, MeasureSpec.UNSPECIFIED)
            val widthMeasureSpec = MeasureSpec.makeMeasureSpec(colspan * stepX, MeasureSpec.UNSPECIFIED)
            view.measure(widthMeasureSpec, heightMeasureSpec)
            val actualWidth = view.measuredWidth
            val actualHeight = view.measuredHeight
            currentOptimalTextSize = (view as ViewWithText).getOptimalTextSize(currentOptimalTextSize, actualWidth - m_GridSpacing, actualHeight - m_GridSpacing)
        }

        //layout
        for (view in m_GameControls[m_GameStage.value]!!) {
            val lp = view.layoutParams as PlanesVerticalLayoutParams
            val rowspan = if (lp.m_RowSpan != 0) lp.m_RowSpan else 1
            val colspan = if (lp.m_ColSpan != 0) lp.m_ColSpan else 1
            val heightMeasureSpec = MeasureSpec.makeMeasureSpec(rowspan * stepY, MeasureSpec.UNSPECIFIED)
            val widthMeasureSpec = MeasureSpec.makeMeasureSpec(colspan * stepX, MeasureSpec.UNSPECIFIED)
            view.measure(widthMeasureSpec, heightMeasureSpec)
            val actualWidth = view.measuredWidth - m_GridSpacing
            val actualHeight = view.measuredHeight - m_GridSpacing
            val viewCenterX = left + lp.m_Col * stepX + colspan * stepX / 2
            val viewCenterY = top + lp.m_Row * stepY + rowspan * stepY / 2
            if (view is ViewWithText) (view as ViewWithText).setTextSize(currentOptimalTextSize)
            view.layout(viewCenterX - actualWidth / 2, viewCenterY - actualHeight / 2, viewCenterX + actualWidth / 2, viewCenterY + actualHeight / 2)

            //view.layout(left + lp.m_Col * stepX, top + lp.m_Row * stepY, left + (lp.m_Col + 1) * stepX, top + (lp.m_Row + 1) * stepY);
        }
    }

    fun setNewRoundStage() {
        m_GameStage = GameStages.GameNotStarted
        invalidate()
        requestLayout()
    }

    fun setGameStage() {
        m_GameStage = GameStages.Game
        invalidate()
        requestLayout()
    }

    fun setBoardEditingStage() {
        m_GameStage = GameStages.BoardEditing
        invalidate()
        requestLayout()
    }

    fun setPlayerBoard() {
        m_ShowPlayerBoard = true
        invalidate()
        requestLayout()
    }

    fun setComputerBoard() {
        m_ShowPlayerBoard = false
        invalidate()
        requestLayout()
    }

    fun setToolbarHeight(toolbarHeight: Int) {
        m_ToolbarHeight = toolbarHeight
    }

    private lateinit var m_GameBoards: ArrayList<GameBoard>
    private lateinit var m_GameControls: HashMap<Int, ArrayList<View>>
    private lateinit var m_GameControlsMaxRow: HashMap<Int, Int>
    private lateinit var m_GameControlsMaxCol: HashMap<Int, Int>
    private var m_ToolbarHeight: Int = 0
    private var m_Context: Context
    private var m_GameStage = GameStages.BoardEditing
    private val m_CorrectChildren = false
    private var m_ShowPlayerBoard = false //in start new game stage which board to show
    private var m_Tablet = false
    private var m_Vertical = false
    private val m_GridSpacing = 5
}